terraform/cache/controllers/downloader.go

```diff
@@ -0,0 +1,65 @@
+package controllers
+
+import (
+	"net/url"
+	"path"
+
+	"github.com/gruntwork-io/terragrunt/pkg/log"
+	"github.com/gruntwork-io/terragrunt/terraform/cache/handlers"
+	"github.com/gruntwork-io/terragrunt/terraform/cache/models"
+	"github.com/gruntwork-io/terragrunt/terraform/cache/router"
+	"github.com/gruntwork-io/terragrunt/terraform/cache/services"
+	"github.com/labstack/echo/v4"
+)
+
+const (
+	downloadPath         = "/downloads"
+	downloadProviderPath = "/provider"
+)
+
+type DownloaderController struct {
+	ReverseProxy    *handlers.ReverseProxy
+	ProviderService *services.ProviderService
+
+	basePath string
+}
+
+// ProviderProxyURL returns URL for using as a cache to download remote archives through this controller with caching
+func (controller *DownloaderController) ProviderProxyURL() *url.URL {
+	cacheURL := *controller.ReverseProxy.ServerURL
+	cacheURL.Path = path.Join(cacheURL.Path, controller.basePath, downloadProviderPath)
+	return &cacheURL
+}
+
+// Register implements router.Controller.Register
+func (controller *DownloaderController) Register(router *router.Router) {
+	router = router.Group(downloadPath)
+	controller.basePath = router.Prefix()
+
+	// Download provider
+	router.GET(downloadProviderPath+"/:remote_host/:remote_path", controller.downloadProviderAction)
+}
+
+func (controller *DownloaderController) downloadProviderAction(ctx echo.Context) error {
+	var (
+		remoteHost = ctx.Param("remote_host")
+		remotePath = ctx.Param("remote_path")
+	)
+
+	provider := &models.Provider{
+		DownloadURL: &url.URL{
+			Scheme: "https",
+			Host:   remoteHost,
+			Path:   "/" + remotePath,
+		},
+	}
+
+	if cache := controller.ProviderService.GetProviderCache(provider); cache != nil {
+		if filename := cache.ArchiveFilename(); filename != "" {
+			log.Debugf("Provider %q uses cached file %q", cache.Provider, filename)
+			return ctx.File(filename)
+		}
+	}
+
+	return controller.ReverseProxy.NewRequest(ctx, provider.DownloadURL)
```

Comment:
Does it make sense to wrap the returned error?

Reply:
This is not the error that is returned to the main function, this error is returned to the HTTP client that made the request for this URL. So we will never see this error in the terminal. Therefore, there is no point in wrapping.